### PR TITLE
Add more descriptive messages when `gem update` fails to update some gems

### PR DESCRIPTION
**Pull request by brianleshopify**
## What was the end-user or developer problem that led to this PR?

When running `gem update` there are points of unclarity when gems are not updated
- it is not clear which gems were not updated
- the reason why gems are not updated (already at latest version or gem isn't even installed)

## What is your fix for the problem, implemented in this PR?

- Add a more verbose message specifying which gems were updated and if some were not, specifying the reason why

Fixes https://github.com/rubygems/rubygems/issues/4306.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Review by deivid-rodriguez — COMMENTED** [review]
Looks very good! Just left a minor code suggestion.

Something else I would do is split the PR into two commits for each improvement.

* Printing the current "already up to date" message regardless of whether any gems were updated, plus the added assertion for that.
* Adding an extra message about gems not updated because of not being installed at all, plus the added test.

File: lib/rubygems/commands/update_command.rb [hunk 1]
```diff
@@ -118,15 +118,19 @@ def execute
...
+    up_to_date_names = not_updated_names - not_installed_names
 
     if updated.empty?
       say "Nothing to update"
     else
       say "Gems updated: #{updated_names.join(' ')}"
-      say "Gems already up-to-date: #{not_updated_names.join(' ')}" unless not_updated_names.empty?
     end
+    say "Gems already up-to-date: #{up_to_date_names.join(' ')}" unless up_to_date_names.empty?
+    say "Gems not installed: #{not_installed_names.join(' ')}" unless not_installed_names.empty?
```
[the remark this answers is not in the document]
Comment (brianleshopify) [reply lib/rubygems/commands/update_command.rb]: Totally agree, I was struggling to think of a phrasing for this

**Review by deivid-rodriguez — COMMENTED** [review]
Just a follow-up to my previous nitpick of splitting commits. 

ec3f4133a643f07b3dbdbff0565228ad05960b73 should move the "Gems already up-to-date" line outside of the if-else block. You added it outside, but left it also inside, which means the intermediate commit does not pass tests because it prints the line twice sometimes. The final diff is the same but I'd like the intermediate commit to also pass tests (it's also handy for bisections).

**Comment by brianleshopify** [comment]
> Just a follow-up to my previous nitpick of splitting commits.
> 
> [ec3f413](https://github.com/rubygems/rubygems/commit/ec3f4133a643f07b3dbdbff0565228ad05960b73) should move the "Gems already up-to-date" line outside of the if-else block. You added it outside, but left it also inside, which means the intermediate commit does not pass tests because it prints the line twice sometimes. The final diff is the same but I'd like the intermediate commit to also pass tests (it's also handy for bisections).

I apologise for that! This was my first time using `git rebase --interactive` to modify my past commits. I must have messed up something while staging.

Thank you for your continued patience and reviews!

**Comment by deivid-rodriguez** [comment]
Thanks for your patience too @brianleshopify, there's just a typo in the second commit message, it should be (I think) something like

> add message when gems are requested to **be** updated **but they are** not installed

Thank you!

**Comment by brianleshopify** [comment]
@deivid-rodriguez Fixed the typo!

**Review by deivid-rodriguez — APPROVED** [review]
Thank you!! 😍

**Pull request merged**